### PR TITLE
Custom Policies for Buckets

### PR DESCRIPTION
**Pull request by adfost**
Allow custom policies for buckets in console UI.

Screen Shot 2021-12-17 at 3 21 09 PM

### CODE REVIEW

File: portal-ui/src/screens/Console/Buckets/BucketDetails/BucketSummaryPanel.tsx [hunk 1]
```diff
@@ -178,9 +178,12 @@ const BucketSummary = ({
   const bucketName = match.params["bucketName"];
 
   let accessPolicy = "n/a";
+  let policyDefinition = "";
 
   if (bucketInfo !== null) {
+    console.log(bucketInfo.definition);
```
Comment (prakashsvmx) [inline portal-ui/src/screens/Console/Buckets/BucketDetails/BucketSummaryPanel.tsx:184]: Please remove the `console.log`

File: portal-ui/src/screens/Console/Buckets/BucketDetails/SetAccessPolicy.tsx [hunk 2]
```diff
@@ -30,6 +30,7 @@ import { ErrorResponseHandler } from "../../../../common/types";
 import api from "../../../../common/api";
 import ModalWrapper from "../../Common/ModalWrapper/ModalWrapper";
 import SelectWrapper from "../../Common/FormComponents/SelectWrapper/SelectWrapper";
+import CodeMirrorWrapper from "../../Common/FormComponents/CodeMirrorWrapper/CodeMirrorWrapper";
 
 const styles = (theme: Theme) =>
```
Comment (prakashsvmx) [inline portal-ui/src/screens/Console/Buckets/BucketDetails/SetAccessPolicy.tsx:35]:
```suggestion
const styles = (theme: Theme) =>
  createStyles({
    codeMirrorContainer: {
      marginBottom: 20,
      paddingLeft: 15,
      "&:nth-child(2) .MuiGrid-root:nth-child(3)": {
        border: "1px solid #EAEAEA",
      },
      "& label": {
        marginBottom: ".5rem",
      },
      "& label + div": {
        display: "none",
      },
    },
    ...modalStyleUtils,
    ...spacingUtils,
  });
```
`codeMirrorContainer`  style needs to be defined. to get a border around it.

**Review by prakashsvmx — CHANGES_REQUESTED** [review]
Setting Custom policy reverts back to the previous once the modal is closed. So Please take a look.

File: restapi/user_buckets.go [hunk 3]
```diff
@@ -436,31 +436,38 @@ func getMakeBucketResponse(session *models.Principal, br *models.MakeBucketReque
...
-	// implemented like minio/mc/ s3Client.SetAccess()
-	if len(bucketAccessPolicy.Statements) == 0 {
-		return client.setBucketPolicyWithContext(ctx, bucketName, "")
-	}
 	policyJSON, err := json.Marshal(bucketAccessPolicy)
 	if err != nil {
 		return err
 	}
-	return client.setBucketPolicyWithContext(ctx, bucketName, string(policyJSON))
+	err = client.setBucketPolicyWithContext(ctx, bucketName, string(policyJSON))
```
Comment (kannappanr) [inline restapi/user_buckets.go:466]: this change is not needed

Comment (adfost) [reply restapi/user_buckets.go]: fixed

**Review by prakashsvmx — DISMISSED** [review]
LGTM.

File: portal-ui/src/screens/Console/Buckets/BucketDetails/SetAccessPolicy.tsx [hunk 4]
```diff
@@ -79,7 +101,8 @@ const SetAccessPolicy = ({
 
   useEffect(() => {
     setAccessPolicy(actualPolicy);
-  }, [setAccessPolicy, actualPolicy]);
+    setPolicyDefinition(actualDefinition);
```
Comment (bexsoft) [inline portal-ui/src/screens/Console/Buckets/BucketDetails/SetAccessPolicy.tsx:104]:
```suggestion
    setPolicyDefinition(actualDefinition);
    setPolicyDefinition(
        actualDefinition ? JSON.stringify(JSON.parse(actualDefinition), null, 4) : ""
      );
```

File: portal-ui/src/screens/Console/Buckets/BucketDetails/SetAccessPolicy.tsx [hunk 5]
```diff
@@ -98,24 +121,34 @@ const SetAccessPolicy = ({
...
+                  { value: "PUBLIC", label: "Public" },
+                  { value: "CUSTOM", label: "Custom" },
+                ]}
+              />
+            </Grid>
+            {accessPolicy === "CUSTOM" && (
+              <Grid item xs={12} className={classes.codeMirrorContainer}>
+                <CodeMirrorWrapper
+                  label={`Write Policy`}
+                  value={policyDefinition}
```
Comment (bexsoft) [inline portal-ui/src/screens/Console/Buckets/BucketDetails/SetAccessPolicy.tsx:145]: Text formatting needs to be applied:

Screen Shot 2021-12-28 at 15 14 47

**Review by bexsoft — CHANGES_REQUESTED** [review]

**Review by bexsoft — APPROVED** [review]
LGTM

**Review by prakashsvmx — APPROVED** [review]

**Pull request merged**